testing/proxytest/proxytest.go

```diff
@@ -176,6 +178,9 @@ func New(t *testing.T, optns ...Option) *Proxy {
 
 	p.Server = httptest.NewUnstartedServer(
 		http.HandlerFunc(func(ww http.ResponseWriter, r *http.Request) {
+			p.requestsWG.Add(1)
```

Comment:
just a maybe stupid question, but my understanding of it is that we increment the waitgroup on each request. Don't we know how many request we are gonna perform on this test so we can add them beforehand and leave only `requestsWG.Done()` inside the handler func?

Reply:
no, we don't. The proxy is used on tests and each test might vary, or even not be possible to know for sure how many requests will go through the proxy

Reply:
just thinking out loud here; 🙂  how do we cover for the scenario where there are still incoming requests but for some reason one of them manages to call Done() on the syncgroup and cause the Wait() in Close() to exit while other incoming ones haven't executed the Add() on the syncgroup yet?!

Reply:
I updated it to close the server first.

Reply:
moving the close before the wait is effectively the same code as before though right?! I mean from the description of the issue "Makes the proxytest to wait all requests to finish before closing the underlying HTTP server" I understand that the reason behind the issue is that we call `p.Server.Close()` before all requests have been done, right?!

Reply:
no, the point is that `p.Server.Close()` should wait all requests to finish, then return. 
The last request log should happen before the server closes, but, as we see in the test failing, some how it isn't the case. 
That's why I added another barrier to try to force the request log to be logged before the test ends.

Reply:
I still don't see how `p.Server.Close()` waits for all requests to finish especially now that it is before `p.requestsWG.Wait()`; `p.Close()` will wait probably for some requests to finish until a `syncgroup.Done()` inside the HandlerFunc manages to lower the syncgroup enough to cause the `p.requestsWG.Wait()` to exit. 🙂

Reply:
That's right, the wait has to come before.
The main idea here is to try to delay a bit the end of the test, just enough for the last request log to be logged before the test end.

That's why it isn't a huge problem if a request slips over the wait group.

Reply:
I am far from an expert in this particular piece of code, but the case that we have here, reminds me a lot some other rare cases where we need close a channel on the receivers end without knowing if the writers are done writing to it. Usually we treat such cases with first shutting down the writers (here should be the components that write to this proxy) then, if we have no shutdown confirmation from the writers, we wait as much as we think the writers should take to shutdown (not ideal) and then we close the channel (the proxy server in this case). Would a similar approach work here?! could we at each test "shutdown" the components that make requests to the proxy first and then close the proxy server?